worker/dbaccessor/tracker.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package dbaccessor
+
+import (
+	"context"
+	"database/sql"
+	"sync"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/worker/v3"
+	"gopkg.in/tomb.v2"
+
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/database"
+)
+
+// TrackedDB defines the union of a TrackedDB and a worker.Worker interface.
+// This is local to the package, allowing for better testing of the underlying
+// trackerDB worker.
+type TrackedDB interface {
+	coredatabase.TrackedDB
+	worker.Worker
+}
+
+// TrackedDBWorkerOption is a function that configures a TrackedDBWorker.
+type TrackedDBWorkerOption func(*trackedDBWorker)
+
+// WithVerifyDBFunc sets the function used to verify the database connection.
+func WithVerifyDBFunc(f func(context.Context, *sql.DB) error) TrackedDBWorkerOption {
+	return func(w *trackedDBWorker) {
+		w.verifyDBFunc = f
+	}
+}
+
+// WithClock sets the clock used by the worker.
+func WithClock(clock clock.Clock) TrackedDBWorkerOption {
+	return func(w *trackedDBWorker) {
+		w.clock = clock
+	}
+}
+
+// WithLogger sets the logger used by the worker.
+func WithLogger(logger Logger) TrackedDBWorkerOption {
```

Comment:
Note that this file is a relocation, not an addition in this patch.

Reply:
sure, I was mostly just curious what you think about the With() pattern, since it was introduced by dqlite, and we're using it in a few more places here. Should we be incorporating this into more of our code?

Reply:
I like this pattern when there is reasonable variability in *behaviours* desired and those are invoked via config. It can be too verbose when there is low variability.

When warranted, I prefer it because it is declarative - semantically composing behaviour rather than just having some config data informing behaviour inside the constructor.